File: lowest-common-ancestor-of-a-binary-search-tree.cpp
//(Learned from discus section)
//Runtime: 26 ms
//Iterative
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        while(1)
        {
            if(root->val > p->val && root->val > q->val)
                root = root->left;
            else if(root->val < p->val && root->val < q->val)
                root = root->right;
            else
                return root;
        }
    }
};

//Runtime: 33 ms
//Recursive
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(p->val < root->val && q->val < root->val)
            return lowestCommonAncestor(root->left, p, q);
        if(p->val > root->val && q->val > root->val)
            return lowestCommonAncestor(root->right, p, q);
        return root;
    }
};

File: lowest-common-ancestor-of-a-binary-search-tree.py
# Runtime: 133 ms


class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        if not root:
            return None
        elif root == p or root == q:
            return root
        elif p.val < root.val and q.val > root.val or p.val > root.val and q.val < root.val:
            return root
        elif root.val < p.val and root.val < q.val:
            return self.lowestCommonAncestor(root.right, p, q)
        else:
            return self.lowestCommonAncestor(root.left, p, q)
